### ISSUE TITLE
Skip dependency of a package, or mark a given dependency as already met

### ISSUE THREAD

**Issue by danielbachhuber** [composer/composer#5024]
I maintain WP-CLI, a command line interface for WordPress similar to Drupal's Drush. In the next release, we're introducing Composer-based package management as a `wp package` command. Here's some of the history: https://github.com/wp-cli/wp-cli/issues/1564

Essentially, here's what the user can do:

* `wp package browse` - Browse all packages available in the Package Index http://wp-cli.org/package-index/
* `wp package list` - List locally-installed packages.
* `wp package install <name>` - Install a package from the Package Index.
* `wp package uninstall <name>` - Uninstall an installed package.

*But*, there's a bit of a problem I've run into (https://github.com/wp-cli/wp-cli/issues/2493). Some of the WP-CLI packages declare `wp-cli/wp-cli` as a dependency. When `wp package install` installs a package with `wp-cli/wp-cli` as a dependency, things generally break. I'd like to skip the `wp-cli/wp-cli` dependency because WP-CLI is already loaded in this context and calling the `vendor/autoload.php` for the locally installed packages.

While I know our use case is atypical and probably falls in the unsupported category, can you think of a reasonable way to skip the `wp-cli/wp-cli` dependency when `Composer\Installer->run()` is executing?

Conceptually, I'd like to indicate the `wp-cli/wp-cli` dependency as already met. But, I haven't found a clean way to do this.

Thanks in advance for your help.

**Comment by curry684** [comment]
I don't know of clean ways to fix this, but there is a nasty workaround that could work for you: the root package name is parsed, not used anywhere in particular, but most certainly injected in the dependency tree.

So if you put this in your root `composer.json`:

    "name": "wp-cli/wp-cli",

It will most certainly consider that package already installed and consider all deeper references to it circular and thus ignore.

**Comment by danielbachhuber** [comment]
> I don't know of clean ways to fix this, but there is a nasty workaround that could work for you: the root package name is parsed, not used anywhere in particular, but most certainly injected in the dependency tree.

Oh, that's an interesting idea. It could actually work fine in our case, as the base project isn't anything more than a shell for the installed packages.

Another suggestion that's come up is to write our own Composer plugin for the installation procedure. It seems a bit heavy, though.

**Comment by danielbachhuber** [comment]
> It could actually work fine in our case, as the base project isn't anything more than a shell for the installed packages.

Lo and behold, it works: https://github.com/wp-cli/wp-cli/issues/2546

@curry684 Can you see this behavior changing anytime soon? If not, I think we can call this idiosyncrasy sufficient for our needs. Conceptually, it mirrors closely to what we want. Given our use case isn't something I think Composer wants to formally support, then I think we can roll with it and see if it causes breakage later.

cc @schlessera

**Comment by curry684** [comment]
It will keep working, because it's a feature on our end, not a bug hehe.

Circular dependencies exist, and sometimes with good reason. So if `foo/bar` depends on `bar/foo` and vice versa, that is a valid situation to us - we just determine install order on first occurrence and skip packages we've already seen when going down the tre.

By that same logic, to be able to develop on either of those packages, we have to inject the root package in the install stack as otherwise it would go all [Inception](http://www.imdb.com/title/tt1375666/) and install itself in itself, possibly even with conflicting versions.

Long story short: the feature is there, feel free to depend on it, I only called it a nasty workaround because you're relying on a side-effect of a wholly different scenario. That's nasty but not forbidden, and unlikely to ever cause issues methinks.

**Comment by danielbachhuber** [comment]
Works for me :) :shipit:

**Comment by danielbachhuber** [comment]
@curry684 Can you think of some way to skip the dependency resolution process for `wp-cli/wp-cli`'s defined dependencies? It's causing a lot of unnecessary memory usage: https://github.com/wp-cli/wp-cli/issues/2566

**Comment by curry684** [comment]
Not really. Recommended practice is still to simply not depend on `wp-cli` from plugins... If package authors need to use classes from your wrapper during development that's understandable, and that's why `require-dev` exists. Dev dependencies are only ever loaded from root packages so that's the one true solution: educating the developers.

The case does pique my interest though as I wouldn't expect the dependency to be traversed at all. Do you have a `-vvv` output to share displaying this observation?

**Comment by danielbachhuber** [comment]
> The case does pique my interest though as I wouldn't expect the dependency to be traversed at all. Do you have a -vvv output to share displaying this observation?

We're actually using Composer as a library, instead of as a binary. Here's the relevant code: https://github.com/wp-cli/wp-cli/blob/master/php/commands/package.php#L67-L127

Our IO class is already set to `isVerbose`, which produces:

```
$ wp package install danielbachhuber/wp-rest-cli
Installing danielbachhuber/wp-rest-cli (dev-master)
Updating /home/vagrant/.wp-cli/packages/composer.json to require the package...
Using Composer to install the package...
---
Loading composer repositories with package information
Updating dependencies
Analyzed 2220 packages to resolve dependencies
Analyzed 29237 rules to resolve dependencies
 - Installing package
Writing lock file
Generating autoload files
---
Success: Package installed successfully.
```

Here's the package in question: https://github.com/danielbachhuber/wp-rest-cli/blob/master/composer.json

Here's the composer.lock for the parent project: https://gist.github.com/danielbachhuber/bc7ccde7f2aaa80a99d8

It doesn't seem that `Composer\Installer` produces much additional output for `isVeryVerbose`, other than:

```
if ($this->verbose && $this->io->isVeryVerbose() && in_array($operation->getJobType(), array('install', 'update'))) {
    $reason = $operation->getReason();
    if ($reason instanceof Rule) {
        switch ($reason->getReason()) {
            case Rule::RULE_JOB_INSTALL:
                $this->io->writeError('    REASON: Required by root: '.$reason->getPrettyString($pool));
                $this->io->writeError('');
                break;
            case Rule::RULE_PACKAGE_REQUIRES:
                $this->io->writeError('    REASON: '.$reason->getPrettyString($pool));
                $this->io->writeError('');
                break;
        }
    }
}
```

**Comment by curry684** [comment]
Walking through the code I think you are doing things right, and I think at https://github.com/wp-cli/wp-cli/blob/master/php/commands/package.php#L238 you are injecting the right parameters, but it would still seem that our code at https://github.com/composer/composer/blob/master/src/Composer/Factory.php#L328 doesn't load the correct package as the root package, or not with the correct data. I would recommend starting the debug there as the `RootPackage` is used to prevent what you are describing from happening, so see if it's correctly loaded with the correct name.

Can't say much more without setting up a full WP-CLI environment myself and debugging into there, and my Wordpress knowledge dates from v1.3 somewhere   :wink:

**Comment by danielbachhuber** [comment]
> I would recommend starting the debug there as the RootPackage is used to prevent what you are describing from happening, so see if it's correctly loaded with the correct name.

Ah, ok. So, just to confirm, based on the information I've provided and the output you see, you think there's a bug? Or at least, some unintentional dependency resolution happening?

**Comment by curry684** [comment]
The only really weird thing I see is:
```
 - Installing package
```
We don't output that string anywhere in that specific form, and we always force a version between parentheses in the places where it does occur. Does your IO class accidentally suppress the fake-HTML markup classes that are in there like `<info>`?

**Comment by danielbachhuber** [comment]
> Does your IO class accidentally suppress the fake-HTML markup classes that are in there like <info>?

Well, somewhat intentionally: https://github.com/wp-cli/wp-cli/blob/master/php/WP_CLI/ComposerIO.php#L34-L37

**Comment by curry684** [comment]
Well `strip_tags` should only strip the actual tags, not their content, so that's not a problem. I just found that message [on your end](https://github.com/wp-cli/wp-cli/blob/b789ef9e03118633d6ca2a1e5482e082ab19775a/php/WP_CLI/PackageManagerEventSubscriber.php#L26), so that explains that.

I also just noticed you're not enabling `--no-dev` in the Installer, that seems weird for a tool meant to install subdependencies. The dependency rules it's resolving now would seem to result from `behat/behat` mostly?

**Comment by danielbachhuber** [comment]
> I also just noticed you're not enabling --no-dev in the Installer, that seems weird for a tool meant to install subdependencies.

`Composer\Installer` defaults to `devMode = false`, so I don't think I'd need to? Explicitly calling `setDevMode( false )` doesn't have an impact on the number of resolved packages.

**Comment by curry684** [comment]
Derp you're right, just about the only place where the default wasn't flipped it seems hehe.
